lib/connection.ts

```diff
@@ -0,0 +1,355 @@
+  session?: Session;
+}
+
+/**
+ * Describes the options that can be provided while creating an AMQP receiver. One can also provide a
+ * session if it was already created.
+ * @interface ReceiverOptionsWithSession
+ */
+export interface ReceiverOptionsWithSession extends ReceiverOptions {
+  session?: Session;
+}
+
+/**
+ * Provides a sender and a receiver link on the same session. It is useful while constructing a
+ * request/response link.
+ *
+ * @interface ReqResLink
+ */
+export interface ReqResLink {
+  /**
+   * @property {Sender} sender - The sender link for sending a request.
+   */
+  sender: Sender;
+  /**
+   * @property {Receiver} receiver - The receiver link for receiving a response.
+   */
+  receiver: Receiver;
+  /**
+   * @property {Session} session - The underlying session on whicn the sender and receiver links
+   * exist.
+   */
+  session: Session;
+}
+
+/**
+ * Descibes the AQMP Connection.
+ * @class Connection
+ */
+export class Connection {
+  /**
+   * @property {rhea.ConnectionOptions} [options] Options that can be provided while creating the
+   * connection.
+   */
+  options?: rhea.ConnectionOptions;
+  /**
+   * @property {rhea.Connection} _connection The connection object from rhea library.
+   * @private
+   */
+  private _connection: rhea.Connection;
+
+  /**
+   * Creates an instance of the Connection object.
+   * @constructor
+   * @param {rhea.Connection} _connection The connection object from rhea library.
+   */
+  constructor(options?: rhea.ConnectionOptions) {
+    this.options = options;
+    this._connection = rhea.create_connection(options);
+  }
+
+  /**
+   * @property {string} id Returns the connection id.
+   * @readonly
+   */
+  get id(): string {
+    return this._connection.options.id!;
+  }
+
+  /**
+   * Creates a new amqp connection.
+   * @param {ConnectionOptions} [options] Options to be provided for establishing an amqp connection.
+   * @return {Promise<Connection>} Promise<Connection>
+   * - **Resolves** the promise with the Connection object when rhea emits the "connection_open" event.
+   * - **Rejects** the promise with an AmqpError when rhea emits the "connection_close" event while trying
+   * to establish an amqp connection.
+   */
+  open(): Promise<Connection> {
+    return new Promise((resolve, reject) => {
+      if (!this.isOpen()) {
+
+        let onOpen: Func<rhea.EventContext, void>;
+        let onClose: Func<rhea.EventContext, void>;
+        let waitTimer: any;
+
+        const removeListeners: Function = () => {
+          clearTimeout(waitTimer);
+          this._connection.removeListener(ConnectionEvents.connectionOpen, onOpen);
+          this._connection.removeListener(ConnectionEvents.connectionClose, onClose);
+          this._connection.removeListener(ConnectionEvents.disconnected, onClose);
+        };
+
+        onOpen = (context: rhea.EventContext) => {
+          removeListeners();
+          process.nextTick(() => {
```

Comment:
why nextTick?

Reply:
also, `process.nextTick()` isn't portable to browsers, so you might be better off with `setTimeout(() => {})` if you want to bump to the next cycle of the loop

Reply:
problem with `setImmediate()` is portability with browsers: https://developer.mozilla.org/en-US/docs/Web/API/Window/setImmediate

Reply:
if setTimeout() is being used then, what is the amount of time, one should wait for? Is 1 second good enough? How much time does a tick take? 1 millisecond?